### PR TITLE
insert compiler libraries head of user-specified classpath - fix for 13552

### PR DESCRIPTION
**Pull request by philwalk**
Reordered `scalaArgs` ahead of `-script <scriptName>` arguments when implementing `ExecuteMode.Script`, as expected by `scripting.Main`.

Expanded argument file, following the pattern in `dotty.tools.dotc.config.CliCommand`.

### CODE REVIEW

File: compiler/src/dotty/tools/MainGenericRunner.scala [hunk 1]
```diff
@@ -137,6 +141,19 @@ object MainGenericRunner {
         val newSettings = if arg.startsWith("-") then settings else settings.withPossibleEntryPaths(arg).withModeShouldBePossibleRun
         process(tail, newSettings.withResidualArgs(arg))
 
+  // copy of method private to dotty.tools.dotc.config.CliCommand.distill()
+  // TODO: make it available as a public method and remove this copy?
```
Comment (smarter) [inline compiler/src/dotty/tools/MainGenericRunner.scala:145]: Yes, please avoid copying code if it can be refactored inside.

**Comment by BarkingBad** [comment]
Maybe we could add some unit tests, for example in `CourisierTests` suite?

File: compiler/src/dotty/tools/MainGenericRunner.scala (hunk 1 above)
Comment (philwalk) [reply compiler/src/dotty/tools/MainGenericRunner.scala:145]: Should I be able to import a method from a trait?   It complains when I try to do so.

If I refactor the `def expandArg` to somewhere outside of `trait CliCommand`, where would a good place be?

**Comment by philwalk** [comment]
> Maybe we could add some unit tests, for example in [CoursierTests](https://github.com/lampepfl/dotty/blob/8d9542cf3b90a5ae0b0a0a85062fa7a0b66277d2/compiler/test-coursier/dotty/tools/coursier/CoursierScalaTests.scala) suite?

I'm working on adding some unit tests.   I ran into a puzzle, though, perhaps someone can suggest what might be going on.
From my build environment, after running `sbt dist/pack`, the following command lines should be equivalent, if not identical:

```bash
bin/scala compiler/test-resources/scripting/classpathReport.sc  # this works as expected
dist/target/pack/bin/scala compiler/test-resources/scripting/classpathReport.sc # this works
compiler/test-resources/scripting/classpathReport.sc # this fails catastrophically in Windows
```

In my environment, calling the script directly prints almost _**half a million compiler error messages**_ (a bit more than 452,000, to be more precise).   Looking at the first few error messages, it seems that the compiler is trying to parse jar files as if they were scala sources:

```scala
[31m-- [E103] Syntax Error: dist\target\pack\lib\antlr-runtime-3.5.1.jar:1:0 -------------------------------------------------------------------------------------------------------------------------------[0m

[31m1 |[0mPK

  |[31m^^^^[0m

  |Illegal start of toplevel definition

longer explanation available when compiling with `-explain`
[31m-- Error: dist\target\pack\lib\antlr-runtime-3.5.1.jar:2:0 ---------------------------------------------------------------------------------------------------------------------------------------------[0m

[31m2 |[0m
```
If the Windows jvm doesn't see a semicolon in a wildcard classpath, it treats it as a glob reference and expands it to a list of files.  See #10761 and #11633.
My hunch is that somehow the wildcard classpath specified in the hashbang is being passed to the jvm unexpanded, although it's not obvious to me why the 3 command lines have differing results.

**Comment by BarkingBad** [comment]
I just tested it on Linux and all 3 commands succeed, however, the last one has many many libraries on classpath, few I recognize from scaladoc

**Comment by philwalk** [comment]
> 
> 
> I just tested it on Linux and all 3 commands succeed, however, the last one has many many libraries on classpath, few I recognize from scaladoc

I'll commit what I have so far, and file a bug report on the bug regression issue.

I added a test to [CoursierScalaTests](https://github.com/lampepfl/dotty/blob/8d9542cf3b90a5ae0b0a0a85062fa7a0b66277d2/compiler/test-coursier/dotty/tools/coursier/CoursierScalaTests.scala) to verify acceptance of an `@argumentFile` on the hashbang line.   The argument file adds `.` (the current directory) to the classpath.   A stronger test would add something not already on the command line and verify that it can be imported, but that will have to wait for later.

**Comment by philwalk** [comment]
Changes in this commit:

- set `script.path` property in `MainGenericRunner` before calling `dotty.tools.scripting.Main`
- append path separator to user classpath from with `dist/bin/scala`
- fix Windows classpath bugs (partial regressions of #10761)

The Windows classpath bug has to be avoided before the jvm is called, since the jvm expands wildcard classpaths before passing control to `MainGenericRunner`.

File: compiler/src/dotty/tools/MainGenericRunner.scala (hunk 1 above)
Comment (smarter) [reply compiler/src/dotty/tools/MainGenericRunner.scala]: > Should I be able to import a method from a trait? It complains when I try to do so.

You could create an object CliCommand in the same file as the trait and put the method there, but CommandLineParser would be fine too.

File: compiler/src/dotty/tools/MainGenericRunner.scala [hunk 2]
```diff
@@ -204,6 +191,8 @@ object MainGenericRunner {
         }
         errorFn("", res)
       case ExecuteMode.Script =>
+        val targetScriptPath: String = settings.targetScript.toString.replace('\\', '/')
+        System.setProperty("script.path", targetScriptPath)
```
Comment (BarkingBad) [inline compiler/src/dotty/tools/MainGenericRunner.scala:195]: Changes are already introduced in scripting [Main](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/scripting/Main.scala). Please decide which place is correct for setting up this property and remove one.

File: compiler/test-coursier/dotty/tools/coursier/CoursierScalaTests.scala [hunk 3]
```diff
@@ -122,6 +122,15 @@ class CoursierScalaTests:
...
 
+    def argumentFile() =
+      // verify that an arguments file is accepted
+      // verify that setting a user classpath does not remove compiler libraries from the classpath.
+      // arguments file contains "-classpath .", adding current directory to classpath.
+      val source = new File(getClass.getResource("/run/myfile.scala").getPath)
+      val argsFile = new File(getClass.getResource("/run/myargs.txt").getPath)
+      val output = CoursierScalaTests.csScalaCmd(s"@$argsFile", source.absPath)
+      assertEquals(output.mkString("\n"), "Hello")
+
```
Comment (BarkingBad) [inline compiler/test-coursier/dotty/tools/coursier/CoursierScalaTests.scala:133]: You forgot to invoke the test by calling `argumentFile()` (like all above functions do). I know that this is not the best way to do it, but JUnit was messing up with starting multiple child processes concurrently, so it has to be done sequentially.

File: dist/bin/scala [hunk 4]
```diff
@@ -28,10 +28,39 @@ fi
...
+    ARGS+=('-classpath')
+    ARGS+=("\"${1#* *}${PSEP}\"")
+    shift
+    ;;
+  *)
+    ARGS+=($1)
+    shift
+    ;;
+  esac
+done
```
Comment (BarkingBad) [inline dist/bin/scala:58]: Is it really necessary to keep it in bash script and not in scala code?

**Review by BarkingBad — CHANGES_REQUESTED** [review]
Good work, but there are some things to fix before merging.

**Comment by BarkingBad** [comment]
To fix test:
in line 158
`val newClasspath = (settings.classPath :+ ".").flatMap(_.split(classpathSeparator).filter(_.nonEmpty)).map(File(_).toURI.toURL)`
in line 180
`val newClasspath = (settings.classPath.flatMap(_.split(classpathSeparator).filter(_.nonEmpty)) ++ removeCompiler(scalaClasspath) :+ ".").map(File(_).toURI.toURL)`

You can create some private function to handle that. The problem is that we try to map to files strings that actually can be enumerated classpath with classpath separators

File: compiler/src/dotty/tools/MainGenericRunner.scala (hunk 2 above)
Comment (philwalk) [reply compiler/src/dotty/tools/MainGenericRunner.scala:195]: Wow, I'm a bit surprised, since I only added it to [MainGenericRunner](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/MainGenericRunner.scala) because a script was unable to see the property.  I'll clean it up.

File: dist/bin/scala (hunk 4 above)
Comment (philwalk) [reply dist/bin/scala:58]: Why do we remove the compiler libraries on line 180?

Comment (BarkingBad) [reply dist/bin/scala:58]: We remove the compiler conditionally. We reuse classpath from coursier, (or script) which bring all the classes on classpath, then filter them out by `removeCompiler` when there is no `-with-compiler` flag. Actually I did a quick look at the issue with the tests and missed the point we gather the classpath. The correct solution should be splitting and flattening the classpath while parsing args at 108 line I guess, so all further processing should have well defined classpath property.

**Comment by philwalk** [comment]
Ouch, I commited bad code (it doesn't compile) ... will commit correct version soon.
Sorry for the noise.

File: compiler/src/dotty/tools/MainGenericRunner.scala (hunk 2 above)
Comment (BarkingBad) [reply compiler/src/dotty/tools/MainGenericRunner.scala]: No problem

File: compiler/src/dotty/tools/MainGenericRunner.scala [hunk 5]
```diff
@@ -102,7 +105,18 @@ object MainGenericRunner {
...
+        (tail, cp)
+      else
+        // combine globbed classpath entries into a classpath
+        val globdir = cp.replaceAll("[\\/][^\\/]*$","") // must be forward-backward-slash-agnostic
+        val jarfiles = cp :: tail
+        val cpfiles = jarfiles.takeWhile( f => f.startsWith(globdir) && ((f.toLowerCase.endsWith(".jar") || f.endsWith(".zip"))) )
+        val tailargs = jarfiles.drop(cpfiles.size)
+        (tailargs, cpfiles.mkString(classpathSeparator))
+        
+      process(tailargs, settings.copy(classPath = settings.classPath.appended(cpstr)))
```
Comment (BarkingBad) [inline compiler/src/dotty/tools/MainGenericRunner.scala:118]:
```suggestion
      process(tailargs, settings.copy(classPath = settings.classPath ++ cpstr.split(classpathSeparator).filter(_.nonEmpty)))
```

We can then revert the changes I proposed previously. Sorry for that.

File: compiler/src/dotty/tools/MainGenericRunner.scala [hunk 6]
```diff
@@ -102,7 +105,18 @@ object MainGenericRunner {
     case "-run" :: fqName :: tail =>
       process(tail, settings.withExecuteMode(ExecuteMode.Run).withTargetToRun(fqName))
     case ("-cp" | "-classpath" | "--class-path") :: cp :: tail =>
-      process(tail, settings.copy(classPath = settings.classPath.appended(cp)))
+      val (tailargs, cpstr) = if classpathSeparator != ";" || cp.contains(classpathSeparator) then
+        (tail, cp)
+      else
+        // combine globbed classpath entries into a classpath
+        val globdir = cp.replaceAll("[\\/][^\\/]*$","") // must be forward-backward-slash-agnostic
+        val jarfiles = cp :: tail
```
Comment (BarkingBad) [inline compiler/src/dotty/tools/MainGenericRunner.scala:113]: Could you check if it is working for `bin/scala -classpath myfile2.jar myfile.jar` (the second jar is with the main class and is supposed to be run). Nonetheless, `bin/scala -classpath myfile2.jar -run myfile.jar` should be completely fine.

File: compiler/src/dotty/tools/MainGenericRunner.scala [hunk 7]
```diff
@@ -187,12 +201,13 @@ object MainGenericRunner {
         }
         errorFn("", res)
       case ExecuteMode.Script =>
+        val targetScriptPath: String = settings.targetScript.toString.replace('\\', '/')
```
Comment (BarkingBad) [inline compiler/src/dotty/tools/MainGenericRunner.scala:204]: Unused value

**Review by BarkingBad — COMMENTED** [review]
Changes look good, just three little nitpicks

File: compiler/src/dotty/tools/MainGenericRunner.scala (hunk 7 above)
Comment (philwalk) [reply compiler/src/dotty/tools/MainGenericRunner.scala:204]: removed

File: compiler/src/dotty/tools/MainGenericRunner.scala (hunk 6 above)
Comment (philwalk) [reply compiler/src/dotty/tools/MainGenericRunner.scala:113]: Wow, cool!  I didn't know you could run a main jar like that (it works for `scala-2.13.6`, but not for `scala3-3.0.2`).
I'll see if we can support it w/out requiring the `-run` arg.

Comment (philwalk) [reply compiler/src/dotty/tools/MainGenericRunner.scala:113]: I assume this is unrelated to this PR, and it's independent of the platform.   The following are from a Linux session.

A difference between `scala3-3.0.2` and the current release candidate is what happens when trying to run a main class, but it's not found in the classpath.    

Here's a command line, intended to run the main method of object invest.Main, if it's not defined in `myfile2.jar`:

```bash
bin/scala -classpath myfile2.jar invest.Main
```
If I misspell the main object, `scala3-3.02` prints an error message:
```scala
Error: Could not find or load main class invest.Main
Caused by: java.lang.ClassNotFoundException: invest.Main
```

When I attempt the same thing with this release candidate, it starts a REPL session, with no error message:
```scala
Welcome to Scala 3.1.1-RC1-bin-SNAPSHOT-git-b9557b0 (16.0.1, Java OpenJDK 64-Bit Server VM).
Type in expressions for evaluation. Or try :help.
                                                                                                                                            
scala> 
```
If I add `-run` ahead of the main object name, it prints the following error:
```scala
philwalk@d5:/opt/ue$ /c/opt/scala3/bin/scala -run invest.Main
Exception in thread "main" java.nio.file.NoSuchFileException: invest.Main
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:219)
	at java.base/java.nio.file.Files.newByteChannel(Files.java:375)
	at java.base/java.nio.file.Files.newByteChannel(Files.java:426)
	at java.base/java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:420)
	at java.base/java.nio.file.Files.newInputStream(Files.java:160)
	at dotty.tools.io.File.inputStream(File.scala:51)
	at dotty.tools.io.Jar.withJarInput(Jar.scala:59)
	at dotty.tools.io.Jar.manifest(Jar.scala:47)
	at dotty.tools.io.Jar.mainClass(Jar.scala:49)
	at dotty.tools.MainGenericRunner$.$anonfun$17(MainGenericRunner.scala:195)
	at scala.Option.flatMap(Option.scala:283)
	at dotty.tools.MainGenericRunner$.run$1(MainGenericRunner.scala:201)
	at dotty.tools.MainGenericRunner$.main(MainGenericRunner.scala:220)
	at dotty.tools.MainGenericRunner.main(MainGenericRunner.scala)
```
The message seems to suggest that it interprets `invest.Main` as a file path, rather than a main object.

Comment (BarkingBad) [reply compiler/src/dotty/tools/MainGenericRunner.scala:113]: That's interesting, I was working on that 2 weeks ago and I thought I covered it with proper testing. Will look into that one tomorrow morning.

Comment (BarkingBad) [reply compiler/src/dotty/tools/MainGenericRunner.scala:113]: Actually, I think this is desired behaviour without `-run` flag, and with `-run` we should catch error and create proper message. The thing is, without `-run` flag we cannot be sure what is the string we are processing - it can be part of some longer flag, like `-mySetting myProperty`. There is heuristic for guessing that string is possibly the main class. If it is not, we assume it's part of residual args. Therefore the REPL start. It's used if somebody want to start REPL with custom configuration.

Comment (BarkingBad) [reply compiler/src/dotty/tools/MainGenericRunner.scala:113]: And for `-run` we firstly try to run it as a main class from classpath, and if it fails we try to run it as a possible runnable file, therefore the error is about `NoSuchFile`, because it fails at fallback. We should create proper message like: The string: <value> is not a valid main class fq name from classpath, nor the valid runnable file.

Comment (BarkingBad) [reply compiler/src/dotty/tools/MainGenericRunner.scala:113]: I know that starting REPL session can be misleading, but I don't know if we could create an efficient parser for all the cases that are: `run` `script` `compile` `repl` that can infer what is the mode that user was thinking about, so we cover happy cases and hope that user knows what he's doing and he won't provide not existing main class. The settings like `-run` and `-script` should help enforcing mode so I think that error messages are the thing we should improve right now.

Comment (philwalk) [reply compiler/src/dotty/tools/MainGenericRunner.scala:113]: Okay, sounds good.

**Comment by philwalk** [comment]
It looks as though we don't check to see if there is a pre-compiled jar  (created via `-save` option from a previous compile), but we always recompile the script source.   The speedup of directly executing a script `.jar` can easily be several orders of magnitude.

The relevant section of [dist/bin/scala](https://github.com/lampepfl/dotty/blob/release-3.0.2/dist/bin/scala) in `scala3-3.0.2` is lines 173-176:
```bash
167 case "${execute_mode-}" in
168 script)
169   if [ "$CLASS_PATH" ]; then
170     script_cp_arg="-classpath '$CLASS_PATH'"
171   fi
172   setScriptName="-Dscript.path=$target_script"
173   target_jar="${target_script%.*}.jar"
174   if [[ $save_compiled == true && "$target_jar" -nt "$target_script" ]]; then
175     eval "\"$JAVACMD\"" $setScriptName -jar "$target_jar" "${script_args[@]}"
176     scala_exit_status=$?
177   else
```

**Comment by philwalk** [comment]
I found and fixed 2 problems, and will commit after some more testing.
In `BashScriptsTests`, there was a platform dependency in a script output validation.
In `settings.scriptArgs`, should have been passed to the precompiled jar instead of `settings.residualArgs`.


For some reason, I can't do the following in my environment, so I have to run the full test suite.
```scala
testOnly dotty.tools.coursier.*
```

**Comment by smarter** [comment]
> For some reason, I can't do the following in my environment, so I have to run the full test suite.

The tests are in their own sbt configuration so you need to run:
```scala
scala3-compiler-bootstrapped/Scala3CompilerCoursierTest/test
```

**Review by BarkingBad — APPROVED** [review]
Looks good

**Comment by philwalk** [comment]
@BarkingBad 
I found a bug related to this PR, and it doesn't yet have test coverage.
Should I create a new PR?

Here's the bug and the one-line fix.

On line 108 of [MainGenericRunner](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/MainGenericRunner.scala):
```scala
      val globdir = cp.replaceAll("[\\/][^\\/]*$", "") // slash/backslash agnostic
```
the regex should have quadruple backslashes, like this:
```scala
      val globdir = cp.replaceAll("[\\\\/][^\\\\/]*$", "") // slash/backslash agnostic
```

The following script fails to compile without the fix:
```scala
#!bin/scala -classpath 'dist/target/pack/lib/*'

// import won't compile unless the hashbang line is effective in setting the classpath
import org.jline.terminal.Terminal

def main(args: Array[String]) =
  val cp = sys.props("java.class.path")
  printf("unglobbed classpath:\n%s\n", cp)
```

**Comment by BarkingBad** [comment]
Yes, fresh PR seems reasonable. Please mention me when you create one

**Comment by philwalk** [comment]
fix and new unit test added by #13619
